x/sync/README.md

```diff
@@ -1,3 +1,57 @@
+
+For each key range, the sync client keeps track of the root ID of the last time it synced that key range. For example, it will store information that says something like, "I have all of the key-value pairs that were in range [`start`, `end`] when the database's root was `rootID`" for some keys `start` and `end`, and some database `rootID`.
+
+If the first range proof contained all of the key-value pairs up to some key `end`, the client recognizes that it must fetch all of the keys after `end`. It divides the remaining key range into chunks, and repeats this process by asking servers for a range proof for each chunk until it has all of the key-value pairs in the database.
+
+However, the database may be changing as the client is syncing. The sync client can be notified that the root ID of the database it's trying to sync to has changed. Detecting that the root ID to sync to has changed is done outside this package. If this occurs, the key-value pairs the client has learned about via range proofs may no longer be valid.
+
+We use change proofs to correct the out of date key-value pairs. When the sync client is notified that the root ID to sync to has changed, it requests a a change proof from a server for a given key range. For example, if a client has the key-value pairs in range [`start`, `end`] that were in the database when it had `rootID`, then it will request a change proof that provides all of the key-value changes in range [`start`, `end`] from the database version with root ID `rootID` to the database version with root ID `newRootID`. The client verifies the change proof, and if it's valid, it applies the changes to its database. If it's not, the client drops the proof and requests the proof from another server.
+
+A server needs to have history in order to serve a change proof. Namely, it needs to know all of the database changes between two roots. If the server does not have sufficient history to generate a change proof, it will send a range proof instead. The client will verify and apply the range proof. Note that change proofs, like range proofs, may not contain all of the key-value pairs in the requested range. This is OK because as mentioned above, the client tracks the root ID associated with each range of key-value pairs it has, so it knows which key-value pairs are out of date. If a change proof doesn't contain all of the key-value pairs in the requested range, the client will request another change proof for the remaining key-value pairs that are out of date.
```

Comment:
might be worth clarifying that it should contain all the changed keys in [start, end'] and will then will request [end', end]

Reply:
done